gateway/handlers/logs.go

```diff
@@ -0,0 +1,133 @@
+package handlers
+
+import (
+	"context"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+)
+
+const crlf = "\r\n"
+const upstreamLogsEndpoint = "/system/logs"
+
+// NewLogHandlerFunc creates and http HandlerFunc from the supplied log Requestor.
+func NewLogHandlerFunc(logProvider url.URL) http.HandlerFunc {
+
+	writeRequestURI := false
+	if _, exists := os.LookupEnv("write_request_uri"); exists {
+		writeRequestURI = exists
+	}
+
+	upstreamLogProviderBase := strings.TrimSuffix(logProvider.String(), "/")
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		if r.Body != nil {
+			defer r.Body.Close()
+		}
+
+		logRequest := buildUpstreamRequest(r, upstreamLogProviderBase, upstreamLogsEndpoint)
+		if logRequest.Body != nil {
+			defer logRequest.Body.Close()
+		}
+
+		hijacker, ok := w.(http.Hijacker)
+		if !ok {
+			log.Println("LogProxy: response is not a Hijacker, required for streaming response")
+			http.NotFound(w, r)
+			return
+		}
+
+		clientConn, buf, err := hijacker.Hijack()
+		if err != nil {
+			log.Println("LogProxy: failed to hijack connection for streaming response")
+			return
+		}
+
+		defer clientConn.Close()
+		clientConn.SetWriteDeadline(time.Time{}) // allow arbitrary time between log writes
+		buf.Flush()                              // will write the headers and the initial 200 response
+
+		if writeRequestURI {
+			log.Printf("LogProxy: proxying request to %s %s\n", logRequest.Host, logRequest.URL.String())
+		}
+
+		ctx, cancel := context.WithCancel(ctx)
+		logRequest = logRequest.WithContext(ctx)
+		defer cancel()
+
+		logResp, err := http.DefaultTransport.RoundTrip(logRequest)
+		if err != nil {
+			log.Printf("LogProxy: forwarding request failed: %s\n", err.Error())
+			buf.WriteString("HTTP/1.1 500 Server Error" + crlf)
+			buf.Flush()
+			return
+		}
+		defer logResp.Body.Close()
+
+		// write response headers directly to the buffer per RFC 2616
+		// https://www.w3.org/Protocols/rfc2616/rfc2616-sec6.html
+		buf.WriteString("HTTP/1.1 " + logResp.Status + crlf)
```

Comment:
I think there's a const for that in the `http` package, not sure if it's public, worth checking.

Reply:
There is no public constant documented in the http package